src/basic/inotify-util.h

```diff
@@ -8,10 +8,16 @@
 
 #define INOTIFY_EVENT_MAX (offsetof(struct inotify_event, name) + NAME_MAX + 1)
 
-#define FOREACH_INOTIFY_EVENT(e, buffer, sz) \
-        for ((e) = &buffer.ev;                                \
-             (uint8_t*) (e) < (uint8_t*) (buffer.raw) + (sz); \
-             (e) = (struct inotify_event*) ((uint8_t*) (e) + sizeof(struct inotify_event) + (e)->len))
+#define _FOREACH_INOTIFY_EVENT(e, buffer, sz, start, end)               \
+        for (struct inotify_event                                       \
+                     *start = &((buffer).ev),                           \
+                     *end = (struct inotify_event*) ((uint8_t*) start + (sz)), \
+                     *e = start;                                        \
+             (uint8_t*) e + sizeof(struct inotify_event) <= (uint8_t*) end && \
+             (uint8_t*) e + sizeof(struct inotify_event) + e->len <= (uint8_t*) end; \
+             e = (struct inotify_event*) ((uint8_t*) e + sizeof(struct inotify_event) + e->len))
```

Comment:
I'm not sure if the added checks are useful. It's part of the inotify api that events in the buffer are delivered atomically, i.e. if `e < start + size`, we can assume that `*e` is a valid event. Maybe this should be an assert instead?

Reply:
`assert()` may be used, but for safety, in the revised version, I added a log message about the invalid event.